master/internal/api_workspace.go

```diff
@@ -12,28 +12,58 @@ import (
 	"google.golang.org/grpc/status"
 
 	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/internal/workspace"
+	"github.com/determined-ai/determined/master/pkg/model"
 	"github.com/determined-ai/determined/proto/pkg/apiv1"
 	"github.com/determined-ai/determined/proto/pkg/workspacev1"
 )
 
-func (a *apiServer) GetWorkspaceByID(id int32, userID int32,
-	rejectImmutable bool,
+func (a *apiServer) GetWorkspaceByID(
+	id int32, curUser model.User, rejectImmutable bool,
 ) (*workspacev1.Workspace, error) {
+	notFoundErr := status.Errorf(codes.NotFound, "workspace (%d) not found", id)
 	w := &workspacev1.Workspace{}
-	switch err := a.m.db.QueryProto("get_workspace", w, id, userID); err {
-	case db.ErrNotFound:
-		return nil, status.Errorf(
-			codes.NotFound, "workspace (%d) not found", id)
-	default:
-		if rejectImmutable && w.Immutable {
-			return nil, errors.Errorf("workspace (%v) is immutable and cannot add new projects.", w.Id)
-		}
-		if rejectImmutable && w.Archived {
-			return nil, errors.Errorf("workspace (%v) is archived and cannot add new projects.", w.Id)
+
+	if err := a.m.db.QueryProto("get_workspace", w, id, curUser.ID); errors.Is(err, db.ErrNotFound) {
+		return nil, notFoundErr
+	} else if err != nil {
+		return nil, errors.Wrapf(err, "error fetching workspace (%d) from database", id)
+	}
+
+	if ok, err := workspace.AuthZProvider.Get().CanGetWorkspace(curUser, w); err != nil {
+		return nil, err
+	} else if !ok {
+		return nil, notFoundErr
```

Comment:
I'm on the fence with this (and CanGetProject) about whether we should specifically acknowledge it's an auth/access issue and not a not-found error. You made a good point earlier that they are sequential so it is not an issue with someone discovering a valid link.

Reply:
The argument for keeping it not found would be consistency and habits in the code base so we don't expose information in cases where it matters more (non sequential IDs).

If we can get to full isolation where a user never sees a project that they don't have access to then a user shouldn't run into it very often? 

I suppose a user could send another user a link to a project like ```127.0.0.1:8080/det/projects/5``` and it might be somewhat surprising to see not found instead of not having access.  

I don't know I'm not really sure which is better to be honest.